src/zjs_net.c

```diff
@@ -0,0 +1,1111 @@
+{
+    sock_handle_t *handle = (sock_handle_t *)h;
+    if (handle) {
+        // find length of unconsumed data in read buffer
+        uint32_t len = handle->wptr - handle->rptr;
+        zjs_buffer_t *zbuf;
+        ZVAL data_buf = zjs_buffer_create(len, &zbuf);
+        // copy data from read buffer
+        memcpy(zbuf->buffer, handle->rptr, len);
+        handle->rptr = handle->wptr = handle->rbuf;
+
+        zjs_trigger_event(handle->socket, "data", &data_buf, 1, NULL, NULL);
+
+        zjs_remove_callback(handle->tcp_read_id);
+    } else {
+        ERR_PRINT("handle is NULL\n");
+    }
+}
+
+static void post_server_closed(void *handle)
+{
+    net_handle_t *h = (net_handle_t *)handle;
+    if (h) {
+        DBG_PRINT("closing server\n");
+        net_context_put(h->tcp_sock);
+        zjs_free(h);
+    }
+}
+
+static void post_closed(void *handle)
+{
+    sock_handle_t *h = (sock_handle_t *)handle;
+    if (h) {
+        net_handle_t *net = h->handle;
+        sock_handle_t *cur = opened_sockets;
+        if (cur->next == NULL) {
+            opened_sockets = NULL;
+            net_context_put(cur->tcp_sock);
+            jerry_release_value(cur->socket);
+            zjs_free(cur->rbuf);
+            zjs_free(cur);
+            DBG_PRINT("Freed socket: opened_sockets=%p\n", opened_sockets);
+        } else {
+            while (cur->next) {
+                if (cur->next == h) {
+                    cur->next = cur->next->next;
+                    net_context_put(cur->tcp_sock);
+                    jerry_release_value(cur->socket);
+                    zjs_free(cur->rbuf);
+                    zjs_free(cur);
+                    DBG_PRINT("Freed socket: opened_sockets=%p\n", opened_sockets);
+                }
+            }
+        }
+        if (net) {
+            if (net->listening == 0 && opened_sockets == NULL) {
+                // no more sockets open and not listening, close server
+                zjs_trigger_event(net->server, "close", NULL, 0, post_server_closed, net);
+                DBG_PRINT("server signaled to close\n");
+            }
+        }
+    }
+}
+
+static void tcp_received(struct net_context *context,
+                         struct net_buf *buf,
+                         int status,
+                         void *user_data)
+{
+
+    sock_handle_t *handle = (sock_handle_t *)user_data;
+
+    if (handle) {
+        start_socket_timeout(handle, handle->timeout);
+
+        uint32_t len = net_nbuf_appdatalen(buf);
+        uint8_t *data = net_nbuf_appdata(buf);
+
+        DBG_PRINT("received data, context=%p, data=%p, len=%u\n", context, data, len);
+        // TODO: Probably not the right way to check if a socket is closed
+        if (len == 0 && data == NULL) {
+            // socket close
+            DBG_PRINT("closing socket, context=%p, socket=%u\n", context, handle->socket);
+            zjs_trigger_event(handle->socket, "close", NULL, 0, post_closed, handle);
```

Comment:
Apparently this *does* mean the socket is closed, so it's not *wrong*. But you still need to call the error event I think, probably *before* the close event?